src/utilities/timeline.ts

```diff
@@ -906,97 +940,192 @@ export function spanInView(span: Span, viewTimeRange: TimeRange) {
+                    filterActivitiesByTime,
+                    spanUtilityMaps,
+                    spansMap,
+                    viewTimeRange,
+                  ),
+                );
+              }
+              items.push({ span });
+            }
+          });
+        }
+        activityNodes.push({
+          activity_type: 'aggregation',
+          children: paginateNodes(children, 'activity', id, discreteTreeExpansionMap),
+          expanded: expanded,
+          id,
+          isLeaf: false,
+          items,
+          label,
+          type: 'Activity',
         });
-      }
-      nodes.push({
-        children: paginateNodes(children, id, activityTreeExpansionMap),
-        expanded: expanded,
-        id,
-        isLeaf: false,
-        items,
-        label,
-        type: 'aggregation',
       });
-    });
-  return nodes;
+  }
+
+  // make the external event subtree
+  const externalEventNodes: DiscreteTreeNode[] = [];
+  if (hasExternalEventsLayer) {
+    if (Object.keys(groupedExternalEvents).length !== 0) {
+      const allKeys = new Set(Object.keys(groupedExternalEvents));
+      // Iterate through all groups - either external event types, external source types, or external sources
+      Array.from(allKeys)
+        .sort()
+        .forEach(type => {
+          const externalEventsGroup = groupedExternalEvents[type];
+          const id = type;
+          const expanded = getNodeExpanded(id, discreteTreeExpansionMap);
+          const label = type;
+          const children: DiscreteTreeNode['children'] = [];
+          const items: DiscreteTreeNode['items'] = [];
+          if (externalEventsGroup) {
+            externalEventsGroup.forEach(externalEvent => {
+              items.push({ externalEvent });
+              children.push({
+                activity_type: undefined, // ignored.
+                children: [],
+                expanded: expanded,
+                id: `${externalEvent.pkey.key}`,
+                isLeaf: true,
+                items: [{ externalEvent }],
+                label: externalEvent.pkey.key,
+                type: 'ExternalEvent',
+              });
+            });
+          }
+          externalEventNodes.push({
+            activity_type: undefined, // ignored.
+            children: paginateNodes(children, 'externalEvent', id, discreteTreeExpansionMap),
+            expanded: expanded,
+            id,
+            isLeaf: false,
+            items: items,
+            label,
+            type: 'ExternalEvent',
+          });
+        });
+    }
+  }
+
+  // if both are present, cluster them
+  if (hasActivityLayer && activityNodes.length && hasExternalEventsLayer && externalEventNodes.length) {
+    const activityAggNode: DiscreteTreeNode = {
+      activity_type: undefined,
+      children: activityNodes,
+      expanded: getNodeExpanded('!!activity-agg', discreteTreeExpansionMap),
+      id: '!!activity-agg',
+      isLeaf: false,
+      items: activityNodes
```

Comment:
This should probably be pulled out into a function
```
function getUniqueNodeItems(nodes: DiscreteTreeNode[]) {
  const uniqueLookup: Set<number> = new Set();
  return nodes
    .flatMap((node: DiscreteTreeNode) => node.items)
    .reduce((flattendNodes: DiscreteTreeNodeItem[], nodeItem: DiscreteTreeNodeItem) => {
      if (nodeItem.directive && !uniqueLookup.has(nodeItem.directive.id)) {
        uniqueLookup.add(nodeItem.directive.id);
        flattendNodes.push(nodeItem);
      }
      return flattendNodes;
    }, []);
}
```

Reply:
Updated to use this in [replace node flattening with more efficient function](https://github.com/NASA-AMMOS/aerie-ui/commit/bc54a3cc1e60f6476cebc17b650ff0af3e9eec3b).